// cmd-ts/dist/cjs/option.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.option = option;
const chalk_1 = __importDefault(require("chalk"));
const Result = __importStar(require("./Result"));
const findOption_1 = require("./newparser/findOption");
const types_1 = require("./types");
function fullOption(config) {
    var _a;
    return {
        description: (_a = config.description) !== null && _a !== void 0 ? _a : config.type.description,
        helpTopics() {
            var _a, _b, _c;
            const displayName = (_a = config.type.displayName) !== null && _a !== void 0 ? _a : "value";
            let usage = `--${config.long}`;
            if (config.short) {
                usage += `, -${config.short}`;
            }
            usage += ` <${displayName}>`;
            const defaults = [];
            if (config.env) {
                const env = process.env[config.env] === undefined
                    ? ""
                    : `=${chalk_1.default.italic(process.env[config.env])}`;
                defaults.push(`env: ${config.env}${env}`);
            }
            if (config.defaultValue) {
                try {
                    const defaultValue = config.defaultValue();
                    if (config.defaultValueIsSerializable) {
                        defaults.push(`default: ${chalk_1.default.italic(defaultValue)}`);
                    }
                    else {
                        defaults.push("optional");
                    }
                }
                catch (e) { }
            }
            else if (config.type.defaultValue) {
                try {
                    const defaultValue = config.type.defaultValue();
                    if (config.type.defaultValueIsSerializable) {
                        defaults.push(`default: ${chalk_1.default.italic(defaultValue)}`);
                    }
                    else {
                        defaults.push("optional");
                    }
                }
                catch (e) { }
            }
            else if (config.onMissing || config.type.onMissing) {
                defaults.push("optional");
            }
            return [
                {
                    category: "options",
                    usage,
                    defaults,
                    description: (_c = (_b = config.description) !== null && _b !== void 0 ? _b : config.type.description) !== null && _c !== void 0 ? _c : "self explanatory",
                },
            ];
        },
        register(opts) {
            opts.forceOptionLongNames.add(config.long);
            if (config.short) {
                opts.forceOptionShortNames.add(config.short);
            }
        },
        async parse({ nodes, visitedNodes, }) {
            const options = (0, findOption_1.findOption)(nodes, {
                longNames: [config.long],
                shortNames: config.short ? [config.short] : [],
            }).filter((x) => !visitedNodes.has(x));
            options.forEach((opt) => visitedNodes.add(opt));
            if (options.length > 1) {
                const error = {
                    message: `Too many times provided. Expected 1, got: ${options.length}`,
                    nodes: options,
                };
                return Result.err({ errors: [error] });
            }
            const valueFromEnv = config.env ? process.env[config.env] : undefined;
            const defaultValueFn = config.defaultValue || config.type.defaultValue;
            const onMissingFn = config.onMissing || config.type.onMissing;
            const option = options[0];
            let rawValue;
            let envPrefix = "";
            if (option === null || option === void 0 ? void 0 : option.value) {
                rawValue = option.value.node.raw;
            }
            else if (valueFromEnv !== undefined) {
                rawValue = valueFromEnv;
                envPrefix = `env[${chalk_1.default.italic(config.env)}]: `;
            }
            else if (defaultValueFn) {
                try {
                    const defaultValue = defaultValueFn();
                    return Result.ok(defaultValue);
                }
                catch (e) {
                    const message = `Default value not found for '--${config.long}': ${e.message}`;
                    return Result.err({
                        errors: [
                            {
                                nodes: [],
                                message,
                            },
                        ],
                    });
                }
            }
            else if (onMissingFn) {
                try {
                    const missingValue = await onMissingFn();
                    return Result.ok(missingValue);
                }
                catch (e) {
                    const message = `Failed to get missing value for '--${config.long}': ${e.message}`;
                    return Result.err({
                        errors: [
                            {
                                nodes: [],
                                message,
                            },
                        ],
                    });
                }
            }
            else {
                // If we reach here, no default or prompt value was available
                const raw = (option === null || option === void 0 ? void 0 : option.type) === "shortOption"
                    ? `-${option === null || option === void 0 ? void 0 : option.key}`
                    : `--${config.long}`;
                return Result.err({
                    errors: [
                        {
                            nodes: options,
                            message: `No value provided for ${raw}`,
                        },
                    ],
                });
            }
            const decoded = await Result.safeAsync(config.type.from(rawValue));
            if (Result.isErr(decoded)) {
                return Result.err({
                    errors: [
                        { nodes: options, message: envPrefix + decoded.error.message },
                    ],
                });
            }
            return Result.ok(decoded.value);
        },
    };
}
function option(config) {
    return fullOption({
        type: types_1.string,
        ...config,
    });
}
