// cmd-ts/dist/cjs/runner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = run;
exports.runSafely = runSafely;
exports.dryRun = dryRun;
exports.parse = parse;
const Result_1 = require("./Result");
const effects_1 = require("./effects");
const errorBox_1 = require("./errorBox");
const parser_1 = require("./newparser/parser");
const tokenizer_1 = require("./newparser/tokenizer");
async function run(ap, strings) {
    const result = await runSafely(ap, strings);
    if ((0, Result_1.isErr)(result)) {
        return result.error.run();
    }
    return result.value;
}
/**
 * Runs a command but does not apply any effect
 */
async function runSafely(ap, strings) {
    const hotPath = [];
    const nodes = parseCommon(ap, strings);
    try {
        const result = await ap.run({ nodes, visitedNodes: new Set(), hotPath });
        if ((0, Result_1.isErr)(result)) {
            throw new effects_1.Exit({
                message: (0, errorBox_1.errorBox)(nodes, result.error.errors, hotPath),
                exitCode: 1,
                into: "stderr",
            });
        }
        return (0, Result_1.ok)(result.value);
    }
    catch (e) {
        if (e instanceof effects_1.Exit) {
            return (0, Result_1.err)(e);
        }
        throw e;
    }
}
/**
 * Run a command but don't quit. Returns an `Result` instead.
 */
async function dryRun(ap, strings) {
    const result = await runSafely(ap, strings);
    if ((0, Result_1.isErr)(result)) {
        return (0, Result_1.err)(result.error.dryRun());
    }
    return result;
}
/**
 * Parse the command as if to run it, but only return the parse result and don't run the command.
 */
function parse(ap, strings) {
    const hotPath = [];
    const nodes = parseCommon(ap, strings);
    return ap.parse({ nodes, visitedNodes: new Set(), hotPath });
}
function parseCommon(ap, strings) {
    const longFlagKeys = new Set();
    const shortFlagKeys = new Set();
    const longOptionKeys = new Set();
    const shortOptionKeys = new Set();
    const registerContext = {
        forceFlagShortNames: shortFlagKeys,
        forceFlagLongNames: longFlagKeys,
        forceOptionShortNames: shortOptionKeys,
        forceOptionLongNames: longOptionKeys,
    };
    ap.register(registerContext);
    const tokens = (0, tokenizer_1.tokenize)(strings);
    return (0, parser_1.parse)(tokens, registerContext);
}
